terragrunt/README.md

```diff
@@ -1,20 +1,122 @@
 
-## Directory structure
+## Directory Structure
 
-The `terragrunt` directory is split into two subdirectories:
-* `modules`: the reuseable module definitions for the logical services we manage.
-* `aws`: AWS account definitions which use the modules in the `modules` directory to configure the various AWS accounts we manage.
+The [Terragrunt] configuration is split into two parts: _modules_ and _state_:
 
-## Running
+- _Modules_ are normal [Terraform] modules that each provision a specific part
+  of our infrastructure. They are stored in the [`modules`](#modules) directory.
+- _State_ represents concrete instances of our infrastructure. For example, the
+  [staging environment for `crates.io`](./accounts/legacy/crates-io-staging) is
+  an instances of the [`crates-io`](./modules/crates-io) [Terraform] module.
+  State is managed using [Terragrunt], and is configured inside the
+  [`accounts`](#accounts) directory.
 
-Running terragrunt requires permissions to the AWS account you are configuring. Assuming you have permission, you can cd into the corresponding service within the `aws` directory and run `terragrunt plan` to see the plan terraform will apply and `terragrunt apply` to actually apply the plan.
+Both concepts are described in more details below.
+
+### `accounts`
+
+The `accounts` subdirectory contains the _state_ of our infrastructure.
+
+A _state_ is a concrete instance of a _service_ or a shared module, which has
+been defined as a [Terraform] module in [`modules`](#modules). States are
+deployed using [Terragrunt], and are grouped by the account in which they run.
+
+Given the following directory structure:
+
+```text
+accounts
+└── dev-desktops-prod
+    ├── westeurope
+    └── westus2
+```
+
+`dev-desktops-prod` is (AWS) _account_ that stores the Terraform state files for
+this environment. `westeurope` and `westus2` are two _states_, i.e. concrete
+instances of the [`dev-desktops-azure`](./modules/dev-desktops-azure) module.
+
+### `modules`
+
+The `modules` subdirectory contains the _configuration_ for our infrastructure.
+We use [Terraform] to declare our infrastructure as code, which we have split
+into composable _modules_.
+
+Modules at the root of the `modules` directory are called _services_ and define
+an app or service within our infrastructure. For example, we have a module for
+our [dev desktops](./modules/dev-desktops-azure) that contains every piece of 
+infrastructure that is required to run one or more instances on Azure.
+
+_Services_ can be deployed multiple times. Each deployment is a concrete
+instance of the configuration, and is managed with [Terragrunt] as a _state_
+inside the [`accounts`](#accounts) directory. For example, `crates.io` has a
+staging and a production _state_, which are both configured through the
+[`crates-io`](./modules/crates-io) _service_ module.
+
+Modules that are generic and can be shared between _services_ are stored in the
```

Comment:
I thought the plan was to get rid of the shared directory. Master does not have a shared directory and I changed #172 to not use one.

Reply:
The reason there is no `modules/shared` directory on `master` is that I thought we were going to use dependencies in Terragrunt when I migrated `crates-io`. But my understanding from our conversation was that we want to use `modules/` for anything that can be a state in Terragrunt, and `modules/shared` for internal dependencies/resources in Terraform.